### PR TITLE
New data source: `azurerm_policy_set_definition`

### LINKED ISSUE #6251: data azurerm_policy_set_definition.html
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

A resource to look up and reference a existing policy definition set 

### New or Affected Resource(s)

* data.azurerm_policy_set_definition

### Potential Terraform Configuration

```
data "azurerm_policy_set_definition" "example" {
  policy_set_definition_id= "<some-guid>"
}
```

### PR DESCRIPTION
**Pull request by ArcturusZhang**
This PR adds data source `azurerm_policy_set_definition`, and also resolves some minor issues in the `policy` service.

Fixes #6251

### CODE REVIEW

File: azurerm/internal/services/policy/data_source_policy_set_definition.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	if managementGroupID != "" {
+		setDefinitions, err = client.ListByManagementGroupComplete(ctx, managementGroupID)
+	} else {
+		setDefinitions, err = client.ListComplete(ctx)
+	}
+	if err != nil {
+		return policy.SetDefinition{}, fmt.Errorf("failed to load Policy Definition List: %+v", err)
+	}
+
+	for setDefinitions.NotDone() {
```
Comment (katbyte) [inline azurerm/internal/services/policy/data_source_policy_set_definition.go:129]: If we found more then 1 result could we return an error?

File: azurerm/internal/services/policy/parse/set_definition.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+type PolicySetDefinitionId struct {
+	Name string
+	PolicyScopeId
+}
+
+// TODO: This parsing function is currently suppressing case difference due to github issue: https://github.com/Azure/azure-rest-api-specs/issues/8353
+func PolicySetDefinitionID(input string) (*PolicySetDefinitionId, error) {
+	// in general, the id of a set definition should be:
+	// {scope}/providers/Microsoft.Authorization/policySetDefinitions/set1
+	regex := regexp.MustCompile(`/providers/[Mm]icrosoft.[Aa]uthorization/policy[Ss]et[Dd]efinitions/`)
```
Comment (katbyte) [inline azurerm/internal/services/policy/parse/set_definition.go:17]: should this be
```suggestion
	regex := regexp.MustCompile(`/providers/[Mm]icrosoft\.[Aa]uthorization/policy[Ss]et[Dd]efinitions/`)
```

File: azurerm/internal/services/policy/tests/data_source_policy_set_definition_test.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+					resource.TestCheckResourceAttr(data.ResourceName, "policy_type", "Custom"),
+					resource.TestCheckResourceAttrSet(data.ResourceName, "parameters"),
+					resource.TestCheckResourceAttrSet(data.ResourceName, "policy_definitions"),
+				),
+			},
+		},
+	})
+}
+
+func testAccDataSourceAzureRMPolicySetDefinition_customByName(data acceptance.TestData) string {
```
Comment (katbyte) [inline azurerm/internal/services/policy/tests/data_source_policy_set_definition_test.go:55]: i'm not sure id we need custom here
```suggestion
func testAccDataSourceAzureRMPolicySetDefinition_byName(data acceptance.TestData) string {
```

File: azurerm/internal/services/policy/tests/data_source_policy_set_definition_test.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	return fmt.Sprintf(`
+%s
+
+data "azurerm_policy_set_definition" "test" {
+  name = azurerm_policy_set_definition.test.name
+}
+`, template)
+}
+
+func testAccDataSourceAzureRMPolicySetDefinition_customByDisplayName(data acceptance.TestData) string {
```
Comment (katbyte) [inline azurerm/internal/services/policy/tests/data_source_policy_set_definition_test.go:66]:
```suggestion
func testAccDataSourceAzureRMPolicySetDefinition_byDisplayName(data acceptance.TestData) string {
```

File: azurerm/internal/services/policy/tests/resource_arm_policy_remediation_test.go [hunk 5]
```diff
@@ -298,7 +298,7 @@ provider "azurerm" {
 data "azurerm_subscription" "current" {}
 
 resource "azurerm_policy_set_definition" "test" {
-  name         = "testPolicySet"
+  name         = "acctest-set-%[1]d"
```
Comment (katbyte) [inline azurerm/internal/services/policy/tests/resource_arm_policy_remediation_test.go:301]:
```suggestion
  name         = "acctest-policy-%[1]d"
```

File: azurerm/internal/services/policy/tests/resource_arm_policy_set_definition_test.go [hunk 6]
```diff
@@ -179,13 +177,12 @@ POLICY_RULE
...
   }
 PARAMETERS
-
 }
 
 resource "azurerm_policy_set_definition" "test" {
   name         = "acctestpolset-%d"
   policy_type  = "Custom"
-  display_name = "acctestpolset-%d"
+  display_name = "acctestpolset-display-%d"
```
Comment (katbyte) [inline azurerm/internal/services/policy/tests/resource_arm_policy_set_definition_test.go:185]: Could we use some uppercase letters here?

File: azurerm/internal/services/policy/tests/resource_arm_policy_set_definition_test.go [hunk 7]
```diff
@@ -179,13 +177,12 @@ POLICY_RULE
     }
   }
 PARAMETERS
-
 }
 
 resource "azurerm_policy_set_definition" "test" {
   name         = "acctestpolset-%d"
```
Comment (katbyte) [inline azurerm/internal/services/policy/tests/resource_arm_policy_set_definition_test.go:183]: and here

File: website/docs/d/policy_set_definition.html.markdown [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+* `management_group_id` - (Optional) Only retrieve Policy Set Definitions from this Management Group.
+
+## Attributes Reference
+
+* `id` - The ID of the Policy Set Definition.
+
+* `description` - The Description of the Policy Set Definition.
+
+* `policy_type` - The Type of the Policy Set Definition. Possible values are `BuiltIn` and `Custom`.
```
Comment (katbyte) [inline website/docs/d/policy_set_definition.html.markdown:45]: We don't need to include possible types as this is a data source
```suggestion
* `policy_type` - The Type of the Policy Set Definition.
```

File: website/docs/d/policy_set_definition.html.markdown [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+* `policy_definitions` - The policy definitions in the policy set definition.
```
Comment (katbyte) [inline website/docs/d/policy_set_definition.html.markdown:47]:
```suggestion
* `policy_definitions` - The policy definitions contained within the policy set definition.
```

File: azurerm/internal/services/policy/data_source_policy_set_definition.go [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+
+			"name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+				ExactlyOneOf: []string{"name", "display_name"},
+			},
+
+			"management_group_id": {
```
[the remark this answers is not in the document]
Comment (ArcturusZhang) [reply azurerm/internal/services/policy/data_source_policy_set_definition.go:41]: Done

File: azurerm/internal/services/policy/data_source_policy_set_definition.go [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+	var err error
+
+	if displayName != "" {
+		setDefinition, err = getPolicySetDefinitionByDisplayName(ctx, client, displayName, managementGroupID)
+		if err != nil {
+			return fmt.Errorf("failed to read Policy Set Definition (Display Name %q): %+v", displayName, err)
+		}
+	}
+	if name != "" {
+		setDefinition, err = getPolicySetDefinition(ctx, client, name, managementGroupID)
```
[the remark this answers is not in the document]
Comment (ArcturusZhang) [reply azurerm/internal/services/policy/data_source_policy_set_definition.go:93]: Done

File: azurerm/internal/services/policy/data_source_policy_set_definition.go [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+	defer cancel()
+
+	name := d.Get("name").(string)
+	displayName := d.Get("display_name").(string)
+	managementGroupID := d.Get("management_group_id").(string)
+
+	var setDefinition policy.SetDefinition
+	var err error
+
+	if displayName != "" {
```
[the remark this answers is not in the document]
Comment (ArcturusZhang) [reply azurerm/internal/services/policy/data_source_policy_set_definition.go:86]: Done

File: azurerm/internal/services/policy/data_source_policy_set_definition.go [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+	definitionBytes, err := json.Marshal(setDefinition.PolicyDefinitions)
+	if err != nil {
+		return fmt.Errorf("unable to flatten JSON for `policy_defintions`: %+v", err)
+	}
+	d.Set("policy_definitions", string(definitionBytes))
+
+	return nil
+}
+
+func getPolicySetDefinitionByDisplayName(ctx context.Context, client *policy.SetDefinitionsClient, displayName, managementGroupID string) (policy.SetDefinition, error) {
```
[the remark this answers is not in the document]
Comment (ArcturusZhang) [reply azurerm/internal/services/policy/data_source_policy_set_definition.go:116]: Done

File: azurerm/internal/services/policy/parse/set_definition.go (hunk 2 above)
Comment (ArcturusZhang) [reply azurerm/internal/services/policy/parse/set_definition.go:17]: Done

File: azurerm/internal/services/policy/data_source_policy_set_definition.go (hunk 1 above)
Comment (ArcturusZhang) [reply azurerm/internal/services/policy/data_source_policy_set_definition.go:129]: Done

File: azurerm/internal/services/policy/tests/data_source_policy_set_definition_test.go (hunk 4 above)
Comment (ArcturusZhang) [reply azurerm/internal/services/policy/tests/data_source_policy_set_definition_test.go:66]: Done

File: azurerm/internal/services/policy/tests/data_source_policy_set_definition_test.go (hunk 3 above)
Comment (ArcturusZhang) [reply azurerm/internal/services/policy/tests/data_source_policy_set_definition_test.go:55]: Done

File: azurerm/internal/services/policy/tests/resource_arm_policy_remediation_test.go (hunk 5 above)
Comment (ArcturusZhang) [reply azurerm/internal/services/policy/tests/resource_arm_policy_remediation_test.go:301]: Done

File: azurerm/internal/services/policy/tests/resource_arm_policy_set_definition_test.go (hunk 6 above)
Comment (ArcturusZhang) [reply azurerm/internal/services/policy/tests/resource_arm_policy_set_definition_test.go:185]: Done

File: azurerm/internal/services/policy/tests/resource_arm_policy_set_definition_test.go (hunk 7 above)
Comment (ArcturusZhang) [reply azurerm/internal/services/policy/tests/resource_arm_policy_set_definition_test.go:183]: Done

File: website/docs/d/policy_set_definition.html.markdown (hunk 8 above)
Comment (ArcturusZhang) [reply website/docs/d/policy_set_definition.html.markdown:45]: Done

File: website/docs/d/policy_set_definition.html.markdown (hunk 9 above)
Comment (ArcturusZhang) [reply website/docs/d/policy_set_definition.html.markdown:47]: Done

**Comment by ArcturusZhang** [comment]
Hi @katbyte I have made some changes according to your comments, please have a look, thanks!

**Comment by ArcturusZhang** [comment]
Hi @katbyte I have resolved the merge conflict, please have a look thanks

**Pull request merged**